File: leetcode/199/rust/solution_2.rs
/*
    199. Binary Tree Right Side View
    Author: Miguel Angel Bermeo Ayerbe
*/
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut queue: VecDeque<(usize, Option<Rc<RefCell<TreeNode>>>)> = VecDeque::new();
        let mut output: Vec<i32> = Vec::new();

        queue.push_back((0, root.clone()));
        while !queue.is_empty() {
            if let Some((level, Some(node_i))) = queue.pop_front() {
                let node_i = node_i.borrow();
                if level == output.len() {
                    output.push(node_i.val);
                } else {
                    output[level] = node_i.val;
                }
                queue.push_back((level + 1, node_i.left.clone()));
                queue.push_back((level + 1, node_i.right.clone()));
            }
        }
        output
    }
}

File: leetcode/1366/c++/solution_2.cpp
/*
    1366. Rank Teams by Votes
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    string rankTeams(vector<string>& votes) {
        std::vector<std::vector<int16_t>> score(27,
                                                std::vector<int16_t>(27, 0));
        for (size_t i = 0; i < votes.size(); ++i) {
            for (size_t j = 0, k = 0; j < votes[i].size(); ++j) {
                k = votes[i][j] - 'A';
                --score[k][j];
                score[k][26] = k;
            }
        }
        std::string output;
        std::sort(score.begin(), score.end());
        size_t len = votes[0].size();
        output.reserve(len);
        for (size_t i = 0; i < len; ++i) {
            output.push_back(score[i][26] + 'A');
        }
        return output;
    }
};